libyul/optimiser/StructuralSimplifier.cpp

```diff
@@ -0,0 +1,121 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <libyul/optimiser/StructuralSimplifier.h>
+#include <libyul/optimiser/Semantics.h>
+#include <libyul/AsmData.h>
+#include <libdevcore/CommonData.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+void StructuralSimplifier::operator()(Block& _block)
+{
+	iterateReplacing(
+		_block.statements,
+		[&](Statement& _stmt) -> boost::optional<vector<Statement>>
+		{
+			if (_stmt.type() == typeid(If))
+			{
+				auto& ifStmt = boost::get<If>(_stmt);
+				if (ifStmt.body.statements.empty())
+				{
+					if (MovableChecker(*ifStmt.condition).movable())
+						return {{}};
+					else
+						return {{ExpressionStatement{ifStmt.location, FunctionalInstruction{
+							ifStmt.location,
+							solidity::Instruction::POP,
+							{std::move(*ifStmt.condition)}
+						}}}};
+				}
+				if (ifStmt.condition->type() == typeid(Literal))
+				{
+					auto const& condition = boost::get<Literal>(*ifStmt.condition);
+					if (isTrue(condition))
+					{
+						(*this)(ifStmt.body);
+						return {{std::move(ifStmt.body.statements)}};
```

Comment:
This seems to require an explicit type for some compilers.

Reply:
Or no - it may just have been one set of brackets too many.